Add support for reference types proposal

aheejin opened the pull request:
This adds support for the [reference type proposal](https://github.com/WebAssembly/reference-types/blob/master/proposals/reference-types/Overview.md). This adds support for
all reference types (`anyref`, `funcref`(=`anyfunc`), and `nullref`) and
four new instructions: `ref.null`, `ref.is_null`, `ref.func`, and new
typed `select`. This also adds subtype relationship support between
reference types.

This does not include table instructions yet. This also does not include
wasm2js support.

Fixes #2444 and fixes #2447.

aheejin commented:
This is NOT ready yet... I'll mark this as non-draft once it is.

aheejin commented:
This is (finally!) ready for review now. The fuzzer runs for a fairly long time, but I'll keep running it in case if there are some remaining issues.

kripken commented on test/binaryen.js/exception-handling.js:
```diff
@@ -18,7 +18,8 @@ function stringify(expr) {
   var module = new Binaryen.Module();
-  module.setFeatures(Binaryen.Features.ExceptionHandling);
+  module.setFeatures(Binaryen.Features.ReferenceTypes |
+                     Binaryen.Features.ExceptionHandling);
```
i wonder if enabling exceptions should automatically enable reference types? not sure what the plan in this area is...

aheejin replied:
Exception handling is dependent on reference types, but we currently have separate flags, like other tools: wabt or v8. I'd like to keep those feature flags separate and not dependent on each other for consistency, and the recommended usage is to control the whole thing using clang's (and emscripten's) `-fwasm-exception` flag, which will [turn on both exception handling and reference type features](https://github.com/llvm/llvm-project/blob/8304781cae97f72c737610460bcb25aa2082807d/clang/lib/Driver/ToolChains/WebAssembly.cpp#L240-L244) and embed the info into the custom section. But here in the test there's no way to turn on both features, so..

kripken commented on test/unit/test_features.py:
```diff
@@ -12,10 +12,12 @@ def check_feature(self, module, error, flag):
-                               input=module, check=False, capture_output=True)
+        p = shared.run_process(
+            shared.WASM_OPT + ['--mvp-features', '--print', '-o', os.devnull] +
+            flag.split(),
```
maybe flag could be an array?

aheejin replied:
Done

aheejin commented on src/binaryen-c.cpp:
```diff
@@ -90,10 +90,7 @@ Literal fromBinaryenLiteral(BinaryenLiteral x) {
-    case Type::exnref: // there's no exnref literals
-    case Type::none:
-    case Type::unreachable:
+    default:
```
Yeah I agree compile time errors are better than runtime ones... I'll add them back in. Maybe we can improve this by using some macro or something and factoring out the type-checking code into the macro.. I'll try that later in a follow-up patch.

aheejin replied:
I just realized this compile time error thing does not work anymore, after we changed `Type` from an enum to a class. So there is no point listing all those types after all. I'll leave them as they are anyway because I already changed them all back. If we want to rely on the compile time error behavior, we have to come up with a new macro or something I guess.

kripken replied:
Oh, does the switch warning not work on the num inside the `Type` class? Seems odd, maybe I'm missing something?

aheejin replied:
I tested this a little, and the difference is this:

```
Type::ValueType type;
switch (type) {
case i32:
  break;
}
```
Here `Type::ValueType` is an enum. This does not compile with `-Werror`, because all other types (i64, f32, f64, ...) are missing in the `switch`. But,

```
Type type;
switch (type) {
case i32:
  break;
}
```
This compiles without warnings, even though there are many types missing in the `switch`. The reason is, here `type` is not considered as an enum. It is a class object, which can be used in `switch` because it has a `constexpr operator uint32_t() const`. So it can be casted to an unsigned int, but it is not an enum, so this switch compiles without warnings. But note that the second usage is far more common in the codebase. (Do we ever explicitly use `Type::ValueType`? I don't think so.)

aheejin commented on CHANGELOG.md:
```diff
@@ -15,6 +15,9 @@ full changeset diff at the end of each section.
 
+- Reference type support is added. Supported instructions are `ref.null`,
+  `ref.is_null`, `ref.func`, and typed `select`. Table instructions are not
+  supported yet.
```
Done. But note that the JS API for switch does not _require_ 'type' parameter; it can still be omitted and people can use it in the old way.

aheejin commented on src/wasm.h:
```diff
@@ -565,10 +568,35 @@ class Expression {
+        return false;
+    }
+  }
+
```
Unless the functions are in this header file, I don't think it will increase the object size. So we can only leave the declarations here and put the definitions in wasm.cpp. Or you don't like the functions are in this `Expression` class, whether the definitions are in this file or not? I think we have three options:
1. Leave the declaration here and put the definition in wasm.cpp
2. Take the function out of `Expression` but still leave the declaration in this file and put the definition in wasm.cpp
3. Move both the declaration and definition to another file (where?)

For 2 there's a downside of cluttering `wasm` namespace, which is effectively the global namespace here. I think 1 should be fine, because if we put the definition to the wasm.cpp, it's not gonna increase the object file size after all. For `canInitializeGlobal`, I don't mind moving `canInitializeGlobal` to `ir/global-utils.h`.

aheejin replied:
We talked in person, and @azakai's concern was not about object file size but he wanted to keep `wasm.h` as minimal as possible, so that it only contains the basic structure of wasm module and people can scan it to understand the structure quickly. I moved those functions to `ir/properties.h`.

We also talked about whether it would be better to use `Const` class for not only `[type].const` instructions but also for `ref.null` and `ref.func` instructions, now that they are constants too. That's a rather big change because `Const` is used everywhere, so we can think about it for a follow-up patch.

aheejin commented on src/wasm.h:
```diff
@@ -565,10 +568,35 @@ class Expression {
     return (const T*)this;
   }
+
+  bool isConstExpression() const {
```
We talked in person whether it would be better to use `Const` class for not only `[type].const` instructions but also for `ref.null` and `ref.func` instructions, now that they are constants too. That's a rather big change because `Const` is used everywhere, so we can think about it for a follow-up patch.  And I renamed `isConstExpression` to `isConstantExpression`, to make it a little less confusing from `Const` expressions.

aheejin commented on src/wasm/wasm-type.cpp:
```diff
@@ -289,6 +299,16 @@ Type getType(unsigned size, bool float_) {
   WASM_UNREACHABLE("invalid size");
 }
 
+bool isLeftSubTypeOfRight(Type left, Type right) {
```
You mean something like this?
```
return getLeastUpperBound(left, right) == right;
```

I don't think I can do this, because `getLeastUpperBound` does special handling for `none` and `unreachable` types. For example, when given `unreachable` and `i32`, `getLeastUpperBound` will return `i32`, but this doesn't mean `unreachable` is a subtype of `i32`. And again, `getLeastUpperBound` returns `none` when given `none` and any other type, but this does not mean those types are subtype of `none`.

The reason `getLeastUpperBound` includes that special handling for `none` and `unreachable` is, this is the way it works when we 'merge' types of different branches targeting a single block. This routine was scattered throughout the code base and now it is done in this single place.

aheejin commented on src/wasm-type.h:
```diff
@@ -119,14 +122,25 @@ constexpr Type i64 = Type::i64;
 unsigned getTypeSize(Type type);
 FeatureSet getFeatures(Type type);
 Type getType(unsigned size, bool float_);
+bool isLeftSubTypeOfRight(Type left, Type right);
```
> can be just `isSubType` - seems natural that it asks if the first param is a subtype of the second, I think?

Done. I named this function `isLeftSubTypeOfRight` because I thought it would be confusing which one is a subtype and which is a supertype... but maybe it'd be not very confusing after all.

> Btw, it's not obvious to me from the name if this is strict or not (is A a subtype of A). Is this obvious in type theory? (in set theory it would probably mean e.g. A is a subset of A). Perhaps a comment is worthwhile.

I think so - I mean, when you say A is a subtype of B, this means `A <: B` or `A <= B`, depending on the notation. To denote a subtype that's not the same as the given type, people usually use 'strict subtype' or something. The [wikipedia article on subtyping](https://en.wikipedia.org/wiki/Subtyping) seems to agree on this. I'll add comment clarifying the subtype here includes the type itself.

> (side issue unrelated to this PR, perhaps these toplevel methods should all be static and on `Type`?)

Done. I also found there are some `Type`-related methods that are outside of `Type` class and moved them into the class in #2556.

aheejin commented on src/ir/manipulation.h:
```diff
@@ -33,14 +33,24 @@ inline OutputType* convert(InputType* input) {
 template<typename InputType> inline Nop* nop(InputType* target) {
-  return convert<InputType, Nop>(target);
+  auto* ret = convert<InputType, Nop>(target);
+  ret->finalize();
```
Done

aheejin replied:
I can't do that, because unlike the simplest ones here (`nop`, `unreachable`, and `ref.null`,  other instructions have fields that have not been assigned at the end of `convert` yet. So calling `finalize` on them causes a segmentation fault.

aheejin commented on src/literal.h:
```diff
@@ -34,6 +35,7 @@ class Literal {
     int32_t i32;
     int64_t i64;
     uint8_t v128[16];
+    const char* func; // function name for funcref
```
There is a [`static_assert`](https://github.com/WebAssembly/binaryen/blob/f2ba91b6340a8c3099ab6a48700c717cbd33599e/src/binaryen-c.cpp#L48-L49) to check if the size of `Literal` and `BinaryenLiteral` are the same. And I wanted to avoid using `Name` in `binaryen-c.h` because this file uses a minimal number of classes from the Binaryen C++ implementation. So I used `const char*` in `BinaryenLiteral`, and to satisfy that `static_assert`, this had to be `const char*` too. (Actually, when I used `Name` in `Literal` and `const char*` in `BinaryenLiteral`, the `static_assert` was OK in Linux, but it failed in Windows. Anyway I think it is safer to use the same set of types in both places.)

kripken replied:
I'm surprised the size is different, as a Name just contains a `char*`? Perhaps windows alignment works differently for classes with a single element?

What worries me about using a `char*` here is the lifetime of the string. For Names we manage that globally.

If we can't use a `Name`, perhaps that could be an internal detail, that is the constructor would get a Name, and `getFunc` would return a `Name`? (a possible downside for the latter is that it might hash the string, when we know that isn't necessary)

aheejin replied:
Now I've changed the `func` in the `Literal` to `Name` in the union and it seems OK. (`BinaryenLiteral` still has `const char*`.) I'm not 100% sure what the previous problem was, but I guess it was when I didn't include it within the union but made it a separate field in `Literal` (and `BinaryenLiteral`), which was when I first implemented reference types' `Literal` differently. So, in short, nevermind. :)

aheejin commented on src/passes/Flatten.cpp:
```diff
@@ -194,8 +196,37 @@ struct Flatten
+            // example,
+            // (block $label0 (result anyref)
+            //   (block $label1 (result nullref)
+            //     (local.tee $0
```
Done. And I realized we don't need `$label1` after all... deleted it.

aheejin commented on src/passes/LocalCSE.cpp:
```diff
@@ -172,9 +172,12 @@ struct LocalCSE : public WalkerPass<LinearExecutionWalker<LocalCSE>> {
   void handle(Expression* curr) {
     if (auto* set = curr->dynCast<LocalSet>()) {
       // Calculate equivalences
+      Function* func = getFunction();
```
Done

aheejin commented on src/passes/MergeLocals.cpp:
```diff
@@ -100,7 +100,8 @@ struct MergeLocals
     }
     // compute all dependencies
-    LocalGraph preGraph(getFunction());
+    Function* func = getFunction();
```
Done

aheejin commented on src/passes/LegalizeJSInterface.cpp:
```diff
@@ -107,14 +107,42 @@ struct LegalizeJSInterface : public Pass {
+      // Gather functions used in 'ref.func'. They should not be removed.
+      std::map<Name, std::atomic<bool>> usedInRefFunc;
+
+      struct RefFuncScanner : public WalkerPass<PostWalker<RefFuncScanner>> {
```
Not really..? `ParallelFunctionAnalysis` produces result of type `<Function*, <Function*, info>>`, so it is useful for a case when each function has different info for each of other functions. This requires result of simpler format, `<Function*, info>`, so I'm not sure if this requires `ParallelFunctionAnalysis`. Please let me know if I'm mistaken.

aheejin commented on src/passes/SimplifyLocals.cpp:
```diff
@@ -577,6 +576,7 @@ struct SimplifyLocals
     this->replaceCurrent(newLocalSet);
     sinkables.clear();
     anotherCycle = true;
+    block->finalize();
```
You mean, like [what the current code does](https://github.com/WebAssembly/binaryen/blob/f2ba91b6340a8c3099ab6a48700c717cbd33599e/src/passes/SimplifyLocals.cpp#L547-L549)? This function (`optimizeBlockReturn`) changes the block type, and this is the code where we recompute the new block type. We were able to compute the block type before by only checking the last argument's type, but now it is not possible because branches targeting this block might have other types, so we have to compute the least upper bound of all branch types + the last element type, which is why I called `block->finalize()` here. Can I do this without calling this again?

aheejin commented on src/tools/execution-results.h:
```diff
@@ -100,17 +104,17 @@ struct ExecutionResults {
       if (results.find(name) == results.end()) {
         std::cout << "[fuzz-exec] missing " << name << '\n';
-        abort();
+        return false;
```
This function is `ExecutionResults::operator==(ExecutionResults& other)`, and I thought the natural thing an `operator==` would do when two things are not equal is to return `false` rather than to abort. And we have [code to abort if the results are not equal](https://github.com/WebAssembly/binaryen/blob/f2ba91b6340a8c3099ab6a48700c717cbd33599e/src/tools/execution-results.h#L92-L95) in the caller side too.

aheejin commented on src/tools/fuzzing.h:
```diff
@@ -901,6 +925,9 @@ class TranslateToFuzzReader {
     if (type == i32 || type == i64) {
       options.add(FeatureSet::Atomics, &Self::makeAtomic);
     }
+    if (type == i32) {
```
Because the return type of `ref.is_null` is `i32`.

aheejin commented on src/tools/fuzzing.h:
```diff
@@ -1093,15 +1120,19 @@ class TranslateToFuzzReader {
   }
 
   Expression* makeIf(Type type) {
+    // ifs cannot have nullref type
```
Oh, nice catch. This is a remnant from the era when `nullref` was not allowed for block/loop/if/types... Deleted it.

aheejin commented on src/tools/fuzzing.h:
```diff
@@ -2180,21 +2190,21 @@ class TranslateToFuzzReader {
 
   Expression* makeSelect(Type type) {
-    return makeDeNanOp(buildSelect({make(i32), make(type), make(type)}));
+    // try to have a subtype for one arm
```
I can make the other arm to be a subtype now with no problem :) I just randomly made only one arm a subtype, but maybe we don't need to do that after all because subtype includes the type itself as well.

aheejin commented on src/wasm/wasm-binary.cpp:
```diff
@@ -1802,11 +1806,16 @@ void WasmBinaryBuilder::processFunctions() {
+    for (auto* ref : refs) {
+      if (auto* call = ref->dynCast<Call>()) {
+        call->target = getFunctionName(index);
+      }
```
Done

aheejin commented on src/wasm-traversal.h:
```diff
@@ -1179,8 +1228,6 @@ struct ExpressionStackWalker : public PostWalker<SubType, VisitorType> {
           return curr;
         }
-      } else {
-        WASM_UNREACHABLE("unexpected expression type");
```
This was a bug. With this, if `expressionStack` contains something other than `Block` of `Loop` the method crashes. I don't think this method was used anywhere before. The current uses of `getBreakTarget` all refer to `ControlFlowWalker::getBreakTarget`.

aheejin commented on src/wasm/wasm-validator.cpp:
```diff
@@ -181,6 +181,29 @@ struct ValidationInfo {
     }
   }
+
+  bool shouldLeftBeSubTypeOfRight(Type left,
```
This was named after `isLeftShouldBeSubTypeOfRight`, but as I renamed it to `Type::isSubType` as you suggested, how about simply `shouldBeSubType`? I'll rename it to `shouldBeSubType` for now. Let me know if you don't like the name.

aheejin commented on src/wasm.h:
```diff
@@ -1070,6 +1099,32 @@ class Pop : public SpecificExpression<Expression::PopId> {
+public:
+  RefIsNull(MixedArena& allocator) {}
+
+  Expression* anyref;
```
Done

aheejin commented:
The previous run of the fuzzer lasted ~87000 iterations, at which point I stopped the execution. Now I've made some changes while addressing the comments, I'll keep running the fuzzer on the new code.

aheejin commented:
Thanks! This depends on #2556, so this can land after that.

aheejin commented:
Ran for more than 100000 iterations, after which it failed due to #2558, so I guess it's pretty safe now.

aheejin commented:
This currently contains diff of #2556 too; I'll rebase this after #2556 lands.

aheejin commented:
I'll merge this, because it's pending for a long time now. If other reviewers have comments, please leave them and I'll address them in follow-up PRs. Thanks!

dschuff commented:
:tada:
